docs/types-and-terminology.md

```diff
@@ -0,0 +1,107 @@
+Types and terminology
+=====================
+
+This is a glossary of the core terms in Redux, along with their type signatures. Types are documented using [Flow notation](http://flowtype.org/docs/quick-reference.html#_).
+
+### State
+
+```js
+type State = any
+```
+
+**State** is a broad term, but in the Redux API it usually refers to the single state value that is managed by the store and returned by `getState()`. It is composed of nested properties and represents the entire state of the Redux store. By convention, the top-level state is an an object or some other key-value collection like a Map, but technically it can be any type.
+
+### Action
+
+```js
+type Action = Object
+```
+
+An **action** is a payload of information that is used to accumulate state. The application uses actions to send data to the store. By convention, they contain a `type` property which indicates the nature of the action being dispatched.
+
+See also **intermediate actions** below.
+
+### Dispatching function
+
+```js
+type Dispatch = (a: Action | IntermediateAction) => any
+```
+
+A **dispatching function** (or simply **dispatch function**) is a function that accepts an action or an intermediate action; it then may or may not dispatch one or more actions to the store.
+
+We must distinguish between dispatching functions in general and the base dispatch function provided by the Store interface. The base dispatch function *always* sends an action to the store's reducer, along with the previous state returned by the store, to calculate a new state. However, the application will rarely call the base dispatch function directly — usually, the base dispatch function is wrapped by middleware. See below for more information.
+
+### Reducer
+
+```js
+type Reducer<S,A> = (state: S, action: A) => S
+```
+
+A **reducer** or **reducing function** is a function that accepts an accumulation and a value and returns a new accumulation. They are used to reduce a collection of values down to a single value. Reducers are not unique to Redux — they are a fundamental concept in functional programming.  Even most non-functional languages, like JavaScript, have a built-in API for reducing. (In JavaScript, it's `Array.prototype.reduce()`.)
```

Comment:
Since functions in JavaScript are closures and first-class citizens it is by all means a functional language by definition. Though it's not considered "pure functional", since functions can have side-effects.